icu4c/source/common/cmemory.h

```diff
@@ -675,6 +679,92 @@ inline H *MaybeStackHeaderAndArray<H, T, stackCapacity>::orphanOrClone(int32_t l
     return p;
 }
 
+/**
+ * A simple memory management class that creates new heap allocated objects (of
+ * any class that has a public constructor), keeps track of them and eventually
+ * deletes them all in its own destructor.
+ *
+ * A typical use-case would be code like this:
+ *
+ *     MemoryPool<CharString> pool;
+ *
+ *     UErrorCode status = U_ZERO_ERROR;
+ *     CharString* s1 = pool.create("Hello", status);
+ *     CharString* s2 = pool.create("World", status);
+ *     if (s1 != nullptr && s2 != nullptr && U_SUCCESS(status)) {
+ *         foo(s1->data(), s2->data());
+ *     }
+ *
+ *     // MemoryPool will take care of deleting the CharString objects.
+ *
+ * It doesn't do anything more than that, and is intentionally kept minimalist.
+ */
+template<typename T>
+class MemoryPool {
+public:
+    MemoryPool() : count(0), pool() {}
+
+    ~MemoryPool() {
+        for (int32_t i = 0; i < count; ++i) {
+            delete pool[i];
+        }
+    }
+
+    MemoryPool(const MemoryPool&) = delete;
+    MemoryPool& operator=(const MemoryPool&) = delete;
+
+    MemoryPool(MemoryPool&& other) U_NOEXCEPT
+            : count(other.count),
+              pool(
+#if U_PLATFORM != U_PF_WINDOWS  // ICU-20235 temporary bug workaround.
+                      std::move(other.pool)
+#else
+                      static_cast<decltype(pool)&&>(other.pool)
+#endif
+              ) {
+        other.count = 0;
+    }
+
+    MemoryPool& operator=(MemoryPool&& other) U_NOEXCEPT {
+        count = other.count;
+        pool =
+#if U_PLATFORM != U_PF_WINDOWS  // ICU-20235 temporary bug workaround.
+                std::move(other.pool);
+#else
+                static_cast<decltype(pool)&&>(other.pool);
+#endif
+        other.count = 0;
+        return *this;
+    }
+
+    /**
+     * Creates a new object of typename T, by forwarding any and all arguments
+     * to the typename T constructor.
+     *
+     * @param args Arguments to be forwarded to the typename T constructor.
+     * @return A pointer to the newly created object, or nullptr on error.
```

Comment:
I don't like having to check each create() return value.

If we can't always have a UErrorCode together with the args-forwarding trick, then I think it would be better to have the following, with which you can have multiple `s1 = pool.adopt(new CharString(..., errorCode), errorCode)` followed by a simple check for `U_FAILURE(errorCode)`:
```
T *adopt(T *p, UErrorCode &errorCode) {
    if (U_FAILURE(errorCode)) {
        delete p;
        return nullptr;
    }
    if (p == nullptr) {
        errorCode = U_MEMORY_ALLOCATION_ERROR;
        return nullptr;
    }
    if (count == pool.getCapacity() && pool.resize(...) == nullptr) {
        delete p;
        errorCode = U_MEMORY_ALLOCATION_ERROR;
        return nullptr;
    }
    return pool[count++] = p;
}
```

Reply:
I think that introducing `UErrorCode` here would be an unnecessary complication. No other code already in `cmemory.h` uses `UErrorCode`, for the only error that's ever reported here is that memory allocation failed and that is most easily reported by just returning `nullptr`. I think it would be a good idea for also this new class to follow that pattern, just like the already existing classes here do.

Let's look at a simple example, calling the default constructor:

```
MemoryPool<CharString> pool;

int* s = pool.create();
if (s != nullptr) // ...
```

Or calling a constructor that doesn't do any `UErrorCode` reporting:

```
MemoryPool<CodePointMatcher> pool;

UChar32 cp = // ...
CodePointMatcher* m = pool.create(cp);
if (m != nullptr) // ...
```

(Both examples above are from actual real code.)

I find the code above simple and intuitive, anyone familiar with C/C++ will immediately understand what it does and why, and it gives all error information that there is to give, that memory allocation failed.

Then contrast that with how calling a constructor that does do `UErrorCode` reporting would look with this API:

```
MemoryPool<CharString> pool;

UErrorCode status = U_ZERO_ERROR;
CharString* s = pool.create("Hello", status);
if (s != nullptr && U_SUCCESS(status)) // ...
```

That isn't really any different from how code that directly allocates such an object using `new` would look like:

```
UErrorCode status = U_ZERO_ERROR;
CharString* s = new CharString("Hello", status);
if (s != nullptr && U_SUCCESS(status)) // ...

delete s;
```

I believe that this similarity, that code calling `MemoryPool::create()` looks just like code calling `new` will make it easy to read and write.

Compare instead with your suggestion to add a `UErrorCode` argument:

```
MemoryPool<CharString> pool;

UErrorCode status = U_ZERO_ERROR;
CharString* s = pool.create(status, "Hello", status);
if (U_SUCCESS(status)) // ...
```

I expect the casual reader to find it non-obvious that the return value is guaranteed to never be `nullptr` when `U_SUCCESS(status)` is `true` and why the same `UErrorCode` needs to be passed twice.

Or with your suggestion to require the caller to invoke `new`:

```
MemoryPool<CharString> pool;

UErrorCode status = U_ZERO_ERROR;
CharString* s = pool.adopt(new CharString("Hello", status), status);
if (U_SUCCESS(status)) // ...
```

That's suddenly a lot more code required than with the API that I'm proposing, has the same issues around return value and double status as the last example, and I also don't think it's self-explanatory what will happen if the `new` call here fails.

I don't think that this would be an improvement over my proposal.

Reply:
The adopt(with UErrorCode) would be use just like any of our factory methods which return nullptr when there is a failure.

However, I am ok with going with your model here.